component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,285 @@
+            this.clean(); // clean before add one element.
+            if (this.cache.size() >= maxSize) {
+                synchronized (this.cache) {
+                    while (this.cache.size() >= maxSize) {
+                        final String keyToRemove = this.cache.keySet().iterator().next();
+                        this.cache.remove(keyToRemove);
+                    }
+                }
+            }
         }
-        return (T) element.value;
+
+        final ScheduledFuture<?> task = timeoutMs > 0 ? this.evictionTask(key, timeoutMs) : null;
+
+        final long endOfValidity = this.calcEndOfValidity(timeoutMs);
+        final ElementImpl element =
+                this.addToMap(key, () -> new ElementImpl(value, toRemove, endOfValidity, task, this.timer));
+
+        return element.getValue(expectedClass);
+    }
+
+    @Override
+    public <T> T computeIfAbsent(final Class<T> expectedClass, final String key, final Predicate<Element> toRemove,
+            final Supplier<T> value) {
+
+        final long timeout = this.getConfigValue(CacheConfiguration::getDefaultEvictionTimeout, -1L);
+        return this.computeIfAbsent(expectedClass, key, toRemove, timeout, value);
+    }
+
+    @Override
+    public <T> T computeIfAbsent(final Class<T> expectedClass,
+            final String key, final long timeoutMs, final Supplier<T> value) {
+        return this.computeIfAbsent(expectedClass, key, null, timeoutMs, value);
+    }
+
+    private ElementImpl addToMap(final String key, final Supplier<ElementImpl> builder) {
+        final String internalKey = internalKey(key);
+        return cache.compute(internalKey, (String k, ElementImpl old) -> //
+        old == null || old.mustBeRemoved() ? builder.get() : old);
+    }
+
+    @Override
+    public <T> T computeIfAbsent(final Class<T> expectedClass, final String key, final Supplier<T> value) {
+        final long timeOut = this.getConfigValue(CacheConfiguration::getDefaultEvictionTimeout, -1L);
+        return computeIfAbsent(expectedClass, key, null, timeOut, value);
+    }
+
+    @PreDestroy
+    public void release() {
+        this.cache.forEach((String k, ElementImpl e) -> e.release());
+        this.cache.clear();
+    }
+
+    private long calcEndOfValidity(final long timeoutMs) {
+        return timeoutMs > 0 ? this.timer.get() + timeoutMs : -1;
     }
 
+
     private String internalKey(final String key) {
         return plugin + '@' + key;
     }
 
-    @AllArgsConstructor
-    public static class Element {
+    public void clean() {
+        Stream<Entry<String, ElementImpl>> elements =  //
+                this.cache.entrySet() //
+                        .stream()     //
+                        .filter(e -> e.getValue().mustBeRemoved());
+
+        final int maxEviction = this.getConfigValue(CacheConfiguration::getMaxDeletionPerEvictionRun, -1);
+        if (maxEviction > 0) {
+            elements = elements.limit(maxEviction);
```

Comment:
If we select a subset of removable items maybe we need to define a sort strategy... Oldest first, nearest timeout or something like that.

Reply:
sort strategy will necessary impact performances O(N.log(N)) vs Constant. As it's 'only' a cache, i prefer keep good performances.

Reply:
Same comment as for the clean and maxsize thread.